src/main/cpp/src/timezones.cu

```diff
@@ -0,0 +1,206 @@
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/lists/list_device_view.cuh>
+#include <cudf/lists/lists_column_device_view.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <thrust/binary_search.h>
+
+#include "timezones.hpp"
+
+using column = cudf::column;
+using column_device_view = cudf::column_device_view;
+using lists_column_device_view = cudf::detail::lists_column_device_view;
+using size_type = cudf::size_type;
+using struct_view = cudf::struct_view;
+
+namespace {
+
+/**
+ * @brief Convert the timestamp value to either UTC or a specified timezone
+ *
+ * This device function uses a binary search to find the instant of the transition
+ * to find the right offset to do the transition.
+ *
+ * To transition to UTC: do a binary search on the tzInstant child column and subtract
+ * the offset
+ *
+ * To transition from UTC: do a binary search on the utcInstant child column and add
+ * the offset
+ *
+ * @tparam typestamp_type type of the input and output timestamp
+ * @param timestamp input timestamp
+ * @param transitions the list column of transitions to figure out the correct offset
+ *                    to adjust the timestamp. The type of the values in this column is
+ *                    LIST<STRUCT<utcInstant: int64, tzInstant: int64, utcOffset: int32>>
+ * @param tz_index the index of the specified zone id in the transitions table
+ * @param to_utc whether we are converting to UTC or converting to the timezone
+ */
+template <typename timestamp_type>
+__device__ timestamp_type convert_timestamp_timezone(timestamp_type const &timestamp,
+                                                     lists_column_device_view const &transitions,
+                                                     size_type tz_index, bool to_utc) {
+
+  using duration_type = typename timestamp_type::duration;
+  using cuda::std::chrono::duration_cast;
+
+  auto epoch_seconds =
+      static_cast<int64_t>(duration_cast<cudf::duration_s>(timestamp.time_since_epoch()).count());
+
+  auto const tz_transitions = cudf::list_device_view{transitions, tz_index};
+
+  auto size = tz_transitions.size();
+
+  cudf::device_span<int64_t const> transition_times(
+      &(transitions.child()
+            .child(to_utc ? 1 : 0)
+            .data<int64_t>()[tz_transitions.element_offset(0)]),
+      static_cast<size_t>(size));
+
+  auto idx = thrust::upper_bound(thrust::seq, transition_times.begin(), transition_times.end(),
+                                 epoch_seconds) -
+             transition_times.begin();
+
+  auto const list_offset = tz_transitions.element_offset(size_type(idx - 1));
+  auto const utc_offset = duration_cast<duration_type>(cudf::duration_s{
+      static_cast<int64_t>(transitions.child().child(2).element<int32_t>(list_offset))});
+  return to_utc ? timestamp - utc_offset : timestamp + utc_offset;
```

Comment:
If `to_utc` were a template parameter, this would read as:
```c++
if constexpr (to_utc) {
  return timestamp - utc_offset;
}
else {
  return timestamp + utc_offset;
}
```
Verbose, but one branch would get eliminated.